profiler/src/ProfilerEngine/Datadog.Profiler.Native/ContentionProvider.cpp

```diff
@@ -112,13 +203,115 @@ void ContentionProvider::OnContention(std::chrono::nanoseconds contentionDuratio
 
     auto [blockingThreadId, blockingThreadName] = currentThreadInfo->SetBlockingThread(0, WStr(""));
-    AddContentionSample(0ns, -1, contentionDuration, blockingThreadId, std::move(blockingThreadName), _emptyStack);
+    AddContentionSample(0ns, -1, WaitType::Lock, contentionDuration, blockingThreadId, std::move(blockingThreadName), _emptyStack);
+}
+
+void ContentionProvider::OnWaitStart(std::chrono::nanoseconds timestamp, uintptr_t associatedObjectId)
+{
+    auto currentThreadInfo = ManagedThreadInfo::CurrentThreadInfo;
+    if (currentThreadInfo == nullptr)
+    {
+        return;
+    }
+
+    // TODO: try to get the type of associatedObjectId to make the difference between Monitor/lock, AutoResetEvent, ManualResetEvent, Mutex and Semaphore
+    // the following code does not work because GetClassFromObject returns CORPROF_E_UNSUPPORTED_CALL_SEQUENCE here
+    //  ClassID classId = 0;
+    //  HRESULT hr = _pCorProfilerInfo->GetClassFromObject(static_cast<ObjectID>(associatedObjectId), &classId);
+    //  if (SUCCEEDED(hr))
+    //  {
+    //      std::string typeName;
+    //      if (_pFrameStore->GetTypeName(classId, typeName))
+    //      {
+    //          std::cout << "WaitStart: " << typeName << std::endl;
+    //      }
+    //  }
+
+    currentThreadInfo->SetWaitStart(timestamp);
+
+    // the MethodTable/ClassID is the first field of the object pointed to by the associatedObjectId
+    if (associatedObjectId != 0)
+    {
+        ClassID classId = *reinterpret_cast<ClassID*>(associatedObjectId);
+
+        std::cout << "   WaitStart(" << std::hex << "0x" << associatedObjectId << ") for classID = 0x" << classId << std::endl;
+
+        currentThreadInfo->SetWaitType(GetWaitType(classId));
+    }
+    else
+    {
+        currentThreadInfo->SetWaitType(WaitType::Unknown);
+        std::cout << "   WaitStart()" << std::endl;
+    }
+}
+
+WaitType ContentionProvider::GetWaitType(ClassID classId)
+{
+    if (classId == _mutexClassID)
+    {
+        return WaitType::Mutex;
+    }
+    else
+    if (classId == _semaphoreClassID)
+    {
+        return WaitType::Semaphore;
+    }
+    else
+    if (classId == _autoResetEventClassID)
+    {
+        return WaitType::AutoResetEvent;
+    }
+    else
+    if (classId == _manualResetEventClassID)
+    {
+        return WaitType::ManualResetEvent;
+    }
+
+    return WaitType::Unknown;
+}
+
+
+void ContentionProvider::OnWaitStop(std::chrono::nanoseconds timestamp)
+{
+    auto currentThreadInfo = ManagedThreadInfo::CurrentThreadInfo;
+    if (currentThreadInfo == nullptr)
+    {
+        return;
+    }
+
+    auto waitStartTimestamp = currentThreadInfo->GetWaitStart();
+    if (waitStartTimestamp == 0ns)
+    {
+        return;
+    }
+
+    auto waitDuration = timestamp - waitStartTimestamp;
+    if (waitDuration < 0ns)
+    {
+        return;
+    }
+    currentThreadInfo->SetWaitStart(0ns);
+
+    // We are not interested in waits that are too short
+    if (waitDuration < 1ms)
```

Comment:
I'm not sure about this. Assuming a memcache request takes 300 us to complete, we would miss all sync-over-async on them, even though it could be a very serious issue.

What would be the issue with not filtering any waits?